aws-samples/amazon-sumerian-hosts
# Host API usability improvements (Babylon)

*Issue #* SUMERIAN-11037

## Description of changes:

### Scope

The focus of this pull request is to encapsulate the ~350 lines of boilerplate code required to instantiate a host in BabylonJS (as demonstrated in our old example HTML file) to improve ease of use for users of the Hosts library. That boilerplate code is now encapsulated in a new module called **HostUtils** which exposes a simple `createHost()` function and a `getCharacterConfig()` function.

The old example app has been removed and two new examples have been added. These are the "hello-world-demo" and "gesture-demo" examples found under `/packages/amazon-sumerian-hosts-babylon/examples`.

#### Out of Scope

- Improving the implementation of the boilerplate code itself.
- Adding a similar HostUtils module to the Three.js implementation. That implementation will continue to use the verbose example file for now.
- Although I've added some minimal unit tests for HostUtils, most of the functionality must be manually tested by running the provided examples and observing expected host behavior.

### Files to review

You can ignore many of the auto-generated files. The files to review explicitly are:

- The following files under `/packages/amazon-sumerian-hosts-babylon/src/Babylon.js/`
  - HostUtils.js
  - index.js
- The new test spec: `/packages/amazon-sumerian-hosts-babylon/test/unit/HostUtils.spec.js`
- All files under `/packages/amazon-sumerian-hosts-babylon/examples`
- Any webpack config files
- Any package.json files
- Review the updated auto-generated API docs in a browser. It is not necessary to review the source code of these autogenerated files. See below for more detail on how to review the updated API docs.

### How to test

> **Prerequisites:**
> You must have a Cognito Identity Pool created in your AWS account. The unauthenticated IAM role for the pool (usually ending in the suffix "Unauth_Role") must have the following managed policies assigned to it:
>
> - AmazonPollyReadOnlyAccess
> - (optional) AmazonLexRunBotsOnly (optional for this PR)

Check out the repository.

In a terminal, `cd` to the repository's root directory.

Run `npm install`.

Update the following file to reflect your Cognito Identity Pool ID: `/packages/amazon-sumerian-hosts-babylon/examples/common/demo-credentials.js`

Run `npm run babylon-examples-serve` from the root of the repo. This will automatically start a web server and open a web browser to a listing of the available demos. Click either of the demos to test. If the hosts work properly, then the code is working!

### How to review API doc updates

Open the following file in a browser: `/docs/index.html`

Use the left-hand nav to navigate to "**Namespaces > babylonjs/HostUtils**".


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

## PR Update 1:
- I've removed generated "docs" files from the repo
